### ISSUE TITLE
Rename via ALTER TABLE throws exception on some tables

### ISSUE THREAD

**Issue by rps-v** [crate/crate#6099]
**CrateDB version**:
2.0.6

**JVM version**:
openjdk version "1.8.0_131"
OpenJDK Runtime Environment (build 1.8.0_131-8u131-b11-0ubuntu1.16.04.2-b11)
OpenJDK 64-Bit Server VM (build 25.131-b11, mixed mode)

**OS version / environment description**:
Ubuntu 16.04.2 LTS (GNU/Linux 4.4.0-47-generic x86_64)

**Problem description**:

Renaming a table with ALTER TABLE query throws an exception on some tables (maybe related to shard count).

**Steps to reproduce**:

Table partitioned on 2 columns, 1434 (configured/started) shards, 1.1GB size, 6.3 million records, 0 replicas.

```sql
alter table "some_data" rename to "some_data_old";
```
```
SQLActionException: INTERNAL_SERVER_ERROR 5000 MultiException in io.crate.executor.transport.ChainableActions$Result.addResultAndError(ChainableActions.java:118)
	at io.crate.executor.transport.ChainableActions$Result.addResultAndError(ChainableActions.java:118)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
	at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2155)
	at io.crate.executor.transport.ChainableActions.rollbackOnErrors(ChainableActions.java:90)
	at io.crate.executor.transport.ChainableActions.lambda$run$1(ChainableActions.java:59)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.crate.action.FutureActionListener.onResponse(FutureActionListener.java:45)
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:91)
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:87)
	at org.elasticsearch.action.support.TransportAction$ResponseFilterChain.proceed(TransportAction.java:224)
	at org.elasticsearch.action.ingest.IngestActionFilter.apply(IngestActionFilter.java:89)
	at org.elasticsearch.action.support.TransportAction$ResponseFilterChain.proceed(TransportAction.java:222)
	at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onResponse(TransportAction.java:249)
	at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onResponse(TransportAction.java:235)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.onResponse(TransportMasterNodeAction.java:152)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.onResponse(TransportMasterNodeAction.java:149)
	at org.elasticsearch.action.admin.indices.open.TransportOpenIndexAction$1.onResponse(TransportOpenIndexAction.java:93)
	at org.elasticsearch.action.admin.indices.open.TransportOpenIndexAction$1.onResponse(TransportOpenIndexAction.java:89)
	at org.elasticsearch.cluster.AckedClusterStateUpdateTask.onAllNodesAcked(AckedClusterStateUpdateTask.java:64)
	at org.elasticsearch.cluster.service.ClusterService$SafeAckedClusterStateTaskListener.onAllNodesAcked(ClusterService.java:874)
	at org.elasticsearch.cluster.service.ClusterService.runTasksForExecutor(ClusterService.java:630)
	at org.elasticsearch.cluster.service.ClusterService$UpdateTask.run(ClusterService.java:915)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:458)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:238)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:201)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Use case**:

**Feature description**:

**Comment by vmishra** [comment]
+1 I have been struggling with this issue too since morning, Even after trying multiple permutations constructs on ALTER TABLE mentioned in the latest documentation, it keeps throwing the exception. Any pointers is appreciated.

```
SQLActionException: INTERNAL_SERVER_ERROR 5000 MultiException in io.crate.executor.transport.ChainableActions$Result.addResultAndError(ChainableActions.java:118)
	at io.crate.executor.transport.ChainableActions$Result.addResultAndError(ChainableActions.java:118)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture.uniHandleStage(CompletableFuture.java:834)
	at java.util.concurrent.CompletableFuture.handle(CompletableFuture.java:2155)
	at io.crate.executor.transport.ChainableActions.rollbackOnErrors(ChainableActions.java:90)
	at io.crate.executor.transport.ChainableActions.lambda$run$1(ChainableActions.java:59)
	at java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:952)
	at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:926)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at io.crate.action.FutureActionListener.onFailure(FutureActionListener.java:53)
	at org.elasticsearch.action.support.TransportAction$1.onFailure(TransportAction.java:97)
	at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onFailure(TransportAction.java:254)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$4.handleException(TransportMasterNodeAction.java:190)
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:957)
	at org.elasticsearch.transport.TcpTransport.lambda$handleException$15(TcpTransport.java:1277)
	at org.elasticsearch.threadpool.ThreadPool.lambda$static$0(ThreadPool.java:147)
	at org.elasticsearch.transport.TcpTransport.handleException(TcpTransport.java:1275)
	at org.elasticsearch.transport.TcpTransport.handlerResponseError(TcpTransport.java:1267)
	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:1211)
	at org.elasticsearch.transport.netty3.Netty3MessageChannelHandler.messageReceived(Netty3MessageChannelHandler.java:73)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)


```

**Comment by mikethebeer** [comment]
@rps-v @vmishra thanks. We have a look on this.

**Comment by mikethebeer** [comment]
@rps-v @vmishra please try to upgrade on version 2.1 since there are some fixes checked-in that might fix your problem. In addition it contains a fix that improved the error message if this exception should occur again. Actually a `MultiException` could be anything here and it's really hard to get a clear picture where the execution fails.

Is it possible that after the upgrade you run the query again and check if the issue still pops up? If yes, plsease attach the stacktrace.

**Comment by rps-v** [comment]
@mikethebeer 

Just tried on 2.1.5 and got this:
```
SQLActionException: INTERNAL_SERVER_ERROR 5000 OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3332)
	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:448)
	at java.lang.StringBuilder.append(StringBuilder.java:136)
	at java.lang.StringBuilder.append(StringBuilder.java:76)
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:484)
	at java.lang.StringBuilder.append(StringBuilder.java:166)
	at java.util.StringJoiner.add(StringJoiner.java:185)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
	at io.crate.exceptions.MultiException.lambda$getMessage$0(MultiException.java:64)
	at io.crate.exceptions.MultiException$$Lambda$2170/1271772846.accept(Unknown Source)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:397)
	at io.crate.exceptions.MultiException.getMessage(MultiException.java:64)
```

Tried with a table with 1.3M records, 504 shards, 245mb size. Tried with heap sizes from 1GB to 4GB and it still fails every time.

**Comment by mikethebeer** [comment]
@rps-v Thanks for the info. We will investigate.

**Comment by mikethebeer** [comment]
@rps-v @vmishra  
- The root cause is already fixed in master, we'll see if we can backport this into 2.1 as well
 - 500 shards is still a lot. The root cause of the issue has to do with the number of shards. Would it be possible to reduce this further? You're less likely to run into the issue with a more reasonable shard number.
 - The `OutOfMemory` exception is actually hiding the root cause. We'll fix that as well (also in 2.1)
